File: java/src/main/java/io/binac/leetcode/ReorderList.java
package io.binac.leetcode;

import io.binac.leetcode.util.ListNode;

/**
 * <p>You are given the head of a singly linked-list. The list can be represented as:</p>
 *
 * <pre>L<sub>0</sub> → L<sub>1</sub> → … → L<sub>n - 1</sub> → L<sub>n</sub>
 * </pre>
 *
 * <p><em>Reorder the list to be on the following form:</em></p>
 *
 * <pre>L<sub>0</sub> → L<sub>n</sub> → L<sub>1</sub> → L<sub>n - 1</sub> → L<sub>2</sub> → L<sub>n - 2</sub> → …
 * </pre>
 *
 * <p>You may not modify the values in the list's nodes. Only nodes themselves may be changed.</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 * <img alt="" src="https://assets.leetcode.com/uploads/2021/03/04/reorder1linked-list.jpg" style="width: 422px; height: 222px;">
 * <pre><strong>Input:</strong> head = [1,2,3,4]
 * <strong>Output:</strong> [1,4,2,3]
 * </pre>
 *
 * <p><strong>Example 2:</strong></p>
 * <img alt="" src="https://assets.leetcode.com/uploads/2021/03/09/reorder2-linked-list.jpg" style="width: 542px; height: 222px;">
 * <pre><strong>Input:</strong> head = [1,2,3,4,5]
 * <strong>Output:</strong> [1,5,2,4,3]
 * </pre>
 *
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li>The number of nodes in the list is in the range <code>[1, 5 * 10<sup>4</sup>]</code>.</li>
 * 	<li><code>1 &lt;= Node.val &lt;= 1000</code></li>
 * </ul>
 */
public class ReorderList {
    public static class Solution1 {
        public void reorderList(ListNode head) {
            if (head == null) return;

            ListNode slow = head, fast = head;
            while (fast != null && fast.next != null) {
                slow = slow.next;
                fast = fast.next.next;
            }

            ListNode dummy = new ListNode(0, slow);
            ListNode p;
            while ((p = slow.next) != null) {
                slow.next = p.next;
                p.next = dummy.next;
                dummy.next = p;
            }

            ListNode p1 = head, p2 = dummy.next, tail = dummy;
            while (p1 != slow) {
                tail = tail.next = p1;
                p1 = p1.next;
                tail = tail.next = p2;
                p2 = p2.next;
            }
        }
    }
}

File: python3/leetcodepy/reorder_list.py
"""
You are given the head of a singly linked-list. The list can be represented as:

L0 → L1 → … → Ln - 1 → Ln

Reorder the list to be on the following form:

L0 → Ln → L1 → Ln - 1 → L2 → Ln - 2 → …

You may not modify the values in the list's nodes. Only nodes themselves may be changed.



Example 1:
https://assets.leetcode.com/uploads/2021/03/04/reorder1linked-list.jpg

Input: head = [1,2,3,4]
Output: [1,4,2,3]

Example 2:
https://assets.leetcode.com/uploads/2021/03/09/reorder2-linked-list.jpg

Input: head = [1,2,3,4,5]
Output: [1,5,2,4,3]


Constraints:

The number of nodes in the list is in the range [1, 5 * 10^4].
1 <= Node.val <= 1000
"""
from typing import Optional

from .utils import ListNode


class Solution1:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head:
            return
        fast, slow = head, head
        while fast is not None and fast.next is not None:
            fast = fast.next.next
            slow = slow.next

        dummy = ListNode(0, slow)
        p = slow.next
        while p is not None:
            slow.next = p.next
            p.next = dummy.next
            dummy.next = p
            p = slow.next

        p1, p2, tail = head, dummy.next, dummy
        while p1 is not slow:
            tail.next = p1
            tail = p1
            p1 = p1.next
            tail.next = p2
            tail = p2
            p2 = p2.next
